File: typo-tester-typo-java/src/java/adventureworks/humanresources/department/DepartmentTest.java
package adventureworks.humanresources.department;

import static org.junit.Assert.*;

import adventureworks.DbNow;
import adventureworks.WithConnection;
import adventureworks.customtypes.Defaulted;
import adventureworks.public_.Name;
import org.junit.Test;

public class DepartmentTest {
  private final DepartmentRepoImpl departmentRepo = new DepartmentRepoImpl();

  @Test
  public void works() {
    WithConnection.run(
        c -> {
          // setup - use short ctor + wither for modifieddate
          var unsaved =
              new DepartmentRowUnsaved(new Name("foo"), new Name("bar"))
                  .withModifieddate(new Defaulted.Provided<>(DbNow.localDateTime()));

          // insert and round trip check
          var saved1 = departmentRepo.insert(unsaved, c);
          var saved2 = unsaved.toRow(saved1::departmentid, saved1::modifieddate);
          assertEquals(saved1, saved2);

          // check field values
          departmentRepo.update(saved1.withName(new Name("baz")), c);
          var all = departmentRepo.selectAll(c);
          assertEquals(1, all.size());
          assertEquals(new Name("baz"), all.get(0).name());

          // delete
          departmentRepo.deleteById(saved1.departmentid(), c);
          var afterDelete = departmentRepo.selectAll(c);
          assertTrue(afterDelete.isEmpty());
        });
  }

  @Test
  public void upsertsWorks() {
    WithConnection.run(
        c -> {
          // setup - use short ctor + wither
          var unsaved =
              new DepartmentRowUnsaved(new Name("foo"), new Name("bar"))
                  .withModifieddate(new Defaulted.Provided<>(DbNow.localDateTime()));

          // Scala: insert and verify upsert
          // inserted2 <- departmentRepo.upsert(saved1.copy(name = newName))
          // _ <- ZIO.succeed(assert(inserted2.rowsUpdated === 1L))
          // saved2 = inserted2.updatedKeys.head
          // _ <- ZIO.succeed(assert(saved2.name === newName))
          //
          // Note: Java's upsert returns Row directly, not UpdateResult with rowsUpdated
          var saved1 = departmentRepo.insert(unsaved, c);
          var newName = new Name("baz");
          var saved2 = departmentRepo.upsert(saved1.withName(newName), c);
          assertEquals(newName, saved2.name());

          // Verify only one row exists (implicit verification that upsert updated, not inserted)
          var all = departmentRepo.selectAll(c);
          assertEquals(1, all.size());
          assertEquals(newName, all.get(0).name());
        });
  }
}

File: typr-tester-typr-scala-old/src/scala/adventureworks/humanresources/department/DepartmentTest.scala
package adventureworks.humanresources.department

import adventureworks.{DbNow, WithConnection}
import adventureworks.customtypes.Defaulted
import adventureworks.public.Name
import org.junit.Assert.*
import org.junit.Test

class DepartmentTest {
  private val departmentRepo = new DepartmentRepoImpl

  @Test
  def works(): Unit = {
    WithConnection {
      val unsaved = DepartmentRowUnsaved(name = Name("foo"), groupname = Name("bar"))
        .copy(modifieddate = Defaulted.Provided(DbNow.localDateTime()))

      val saved1 = departmentRepo.insert(unsaved)
      val saved2 = unsaved.toRow(saved1.departmentid, saved1.modifieddate)
      assertEquals(saved1, saved2)

      val _ = departmentRepo.update(saved1.copy(name = Name("baz")))
      val all = departmentRepo.selectAll
      assertEquals(1, all.size)
      assertEquals(Name("baz"), all.get(0).name)

      val _ = departmentRepo.deleteById(saved1.departmentid)
      val afterDelete = departmentRepo.selectAll
      assertTrue(afterDelete.isEmpty)
    }
  }

  @Test
  def upsertsWorks(): Unit = {
    WithConnection {
      val unsaved = DepartmentRowUnsaved(name = Name("foo"), groupname = Name("bar"))
        .copy(modifieddate = Defaulted.Provided(DbNow.localDateTime()))

      val saved1 = departmentRepo.insert(unsaved)
      val newName = Name("baz")
      val saved2 = departmentRepo.upsert(saved1.copy(name = newName))
      assertEquals(newName, saved2.name)

      val all = departmentRepo.selectAll
      assertEquals(1, all.size)
      assertEquals(newName, all.get(0).name)
    }
  }
}
